parser/impl/parser.cpp

```diff
@@ -7,619 +7,664 @@ namespace parser {
     }
-    currentToken = skipWhile(isNewLine);
 
-    if (currentToken.type != TokenType::Is) {
-        m_errors.push_back(Error{
-            .pos = currentToken.pos,
-            .message = "Expected to find keyword 'is'",
-        });
-        // skip till "end" keyword
-        while (m_lexer.Next().type != TokenType::End)
-            ;
+    if (currentToken.type == TokenType::Illegal) {
+        advance(TokenType::End);
         return nullptr;
     }
-    while (m_lexer.Peek().type != TokenType::NewLine)
-        m_lexer.Next();
+
+    skipWhitespace();
+
     routineNode.body = parseBody();
-    currentToken = skipWhile(isNewLine);
-    if (currentToken.type != TokenType::End) {
-        m_errors.push_back(Error{
-            .pos = currentToken.pos,
-            .message = "Expected to find keyword 'end'",
-        });
-        // skip till "end" keyword
-        while (m_lexer.Next().type != TokenType::End)
-            ;
+    currentToken = expect(TokenType::End, "Expected \"end\" keyword.");
+    if (currentToken.type == TokenType::Illegal) {
+        advance(TokenType::End);
         return nullptr;
     }
-    return std::make_shared<ast::Routine>(routineNode);
+    routineNode.end = currentToken.pos;
+    return std::make_shared<ast::RoutineDecl>(routineNode);
 }
 
 sPtr<ast::Parameter> Parser::parseParameter() {
     ast::Parameter parameterNode;
-    Token currentToken = skipWhile(isNewLine);
-    if (currentToken.type != TokenType::Identifier) {
-        m_errors.push_back(Error{
-            .pos = currentToken.pos,
-            .message = "Expected to find an identifier",
-        });
-        // skip till ')' or ','
-        while (m_lexer.Peek().type != TokenType::Comma &&
-               m_lexer.Peek().type != TokenType::CloseParen) {
-            m_lexer.Next();
-        }
+    Token currentToken =
+        expect(TokenType::Identifier, "Expected to find an identifier.");
+    if (currentToken.type == TokenType::Illegal) {
+        advance({TokenType::CloseParen, TokenType::Comma, TokenType::NewLine});
         return nullptr;
     }
-    parameterNode.name = currentToken;
-    currentToken = skipWhile(isNewLine);
-    if (currentToken.type != TokenType::Colon) {
-        m_errors.push_back(Error{
-            .pos = currentToken.pos,
-            .message = "Expected to find a ':'",
-        });
-        // skip till ')' or ','
-        while (m_lexer.Peek().type != TokenType::Comma &&
-               m_lexer.Peek().type != TokenType::CloseParen) {
-            m_lexer.Next();
-        }
+    parameterNode.begin = currentToken.pos;
+    parameterNode.name = currentToken.lit;
+
+    currentToken = expect(TokenType::Colon, "Expected to find a ':'.");
+    if (currentToken.type == TokenType::Illegal) {
+        // Note: the following is not correct as it will consume the ) or ,
+        //  token expected by `parseRoutineDecl`. A better alternative is needed
```

Comment:
Is this comment still valid? Which examples make it fail?

Reply:
Yes, it is valid. Consider `routine main(x integer): integer` ; it will not find the colon between the parameter name and its type, so it will report an error and advance to the first occurrence of `(`/`\n`/`,` and consume it, although `parseRoutineDecl` should be the one to consume the `(`/`,` (line 82).
This comment is to remind us in the future that this can be a source of error, but it doesn't matter much since all it causes to happen is another error (a generic `"Unexpected token"`) be reported